src/typeahead/index.tsx

```diff
@@ -0,0 +1,319 @@
+	initialValue?: string;
+	/** Property to determine how many items to render. Defaults to 6 */
+	itemsInView?: number;
+	/** Placement of the select menu; 'above' or 'below' */
+	position?: PopupPosition;
+	/** Property to determine if the input is disabled */
+	disabled?: boolean;
+	/** Sets the helper text of the input */
+	helperText?: string;
+	/** The label to show */
+	label?: RenderResult;
+	/** Boolean to indicate if field is required */
+	required?: boolean;
+	/** Callback when valid state has changed */
+	onValidate?(valid?: boolean): void;
+	/** The name property of the input */
+	name?: string;
+}
+
+export interface TypeaheadICache {
+	value: string;
+	activeIndex: number;
+	dirty: boolean;
+	expanded: boolean;
+	focusNode: string;
+	initial: string;
+	valid: boolean;
+}
+
+export interface SelectChildren {
+	/** Custom renderer for item contents */
+	(properties: ItemRendererProperties): RenderResult;
+}
+
+const factory = create({
+	icache: createICacheMiddleware<TypeaheadICache>(),
+	data: createDataMiddleware<ListOption>(),
+	theme,
+	focus,
+	i18n
+})
+	.properties<TypeaheadProperties>()
+	.children<SelectChildren | undefined>();
+
+export const Typeahead = factory(function Typeahead({
+	id,
+	properties,
+	children,
+	middleware: { icache, data, theme, focus, i18n }
+}) {
+	const {
+		initialValue,
+		disabled,
+		label,
+		required,
+		position,
+		name,
+		helperText,
+		itemsInView,
+		transform,
+		onValidate
+	} = properties();
+	const themedCss = theme.classes(css);
+	const { messages } = i18n.localize(bundle);
+	const { get, getOptions, setOptions, shared, getTotal, isLoading } = data();
+	const sharedResource = shared();
+
+	const [itemRenderer] = children();
+
+	if (initialValue !== undefined && initialValue !== icache.get('initial')) {
+		icache.set('initial', initialValue);
+		icache.set('value', initialValue);
+	}
+
+	let valid = icache.get('valid');
+	const value = icache.get('value');
+	const listId = `typeahead-list-${id}`;
+	const triggerId = `typeahead-trigger-${id}`;
+	const dirty = icache.get('dirty');
+
+	if (required && dirty) {
+		const isValid = Boolean(value);
+		if (isValid !== valid) {
+			icache.set('valid', isValid);
+			valid = isValid;
+			onValidate && onValidate(isValid);
+		}
+	}
+
+	function onKeyDown(
+		event: number,
+		preventDefault: () => void,
+		onOpen: () => boolean,
+		onClose: () => void
+	) {
+		const activeIndex = icache.getOrSet('activeIndex', 0);
+		const total = getTotal(getOptions()) || 0;
+
+		switch (event) {
+			case Keys.Escape:
+				onClose();
+				break;
+			case Keys.Down:
+				preventDefault();
+				if (!onOpen()) {
+					icache.set('activeIndex', (activeIndex + 1) % total);
+				}
+				break;
+			case Keys.Up:
+				preventDefault();
+				if (!onOpen()) {
+					icache.set('activeIndex', (activeIndex - 1 + total) % total);
+				}
+				break;
+			case Keys.Enter:
+				preventDefault();
+
+				const allItems = get({ query: getOptions().query }) as ListOption[];
```

Comment:
is the `get` result not already typed to `ListOption` due to the the generic passed when the middleware was created?